### PR TITLE
Add shipping restriction details hover

### LINKED ISSUE #729: Product Page: Implement shipping restriction details hover
image

Data will be provided in a variant metafield `shipping_restrictions`, which is an array of codes like `is_battery` or `us_only`.

Instead of displaying "Shipping restrictions apply" in all cases, we'll have a short `notice` and a long `text` (that shows in a tooltip) for each restriction, though some of these do have "Shipping restrictions apply" as their notice. We'll only ever show one restriction, for now, even though we're including an array of all of them. The array has been pre-sorted to have the most relevant first.

When a "link" is specified for a given restriction, add a "Learn more" link at the bottom of the tooltip as shown here:
image
(https://www.ifixit.com/Store/Tools/iOpener/IF145-198?o=9)
For reference, the priority order is:

- battery
- dangerous
- us only
- does not ship to mexico

But this shouldn't need to be coded into the frontend.

Additionally, when a product is fulfilled via dropshipping, we'll include the 'fulfilled_via_dropshipping' key in shipping_restrictions. In this case, we'll show an identical (additional) indicator with the fulfilled_via_dropshipping message specified below. This is the only case where we'll show multiple "shipping restrictions".

Currently this is implemented in ifixit/ifixit: carpenter-frontend/Cart/Product/shipping_flag_notice.js . We should be mostly copy/paste this.

```js
 const attributeToInfo = {
    is_battery: {
       notice: 'Shipping restrictions apply',
       text: 'Batteries may only be shipped within the contiguous USA at this time and may only ship via standard shipping.',
       link: 'https://help.ifixit.com/article/195-battery-shipping-warnings',
    },
    is_dangerous: {
       notice: 'Shipping restrictions apply',
       text: 'New safety regulations for dangerous goods may cause shipping delays.',
       link: 'https://help.ifixit.com/article/264-what-are-the-shipping-restrictions-for-orders-containing-dangerous-goods',
    },
    us_only: {
       notice: 'US shipping only',
       text: 'Due to shipping regulations, iFixit may only ship this product within the United States of America.',
    },
    does_not_ship_to_mexico: {
       notice: 'Mexico shipping unavailable',
       text: 'Due to shipping regulations, iFixit may not ship this product to Mexico.',
    },
    fulfilled_via_dropshipping: {
       notice: 'Partner fulfilled',
       text: 'This product will be packed and shipped by one of our fulfillment partners.',
    },
 };
```

### PR DESCRIPTION
**Pull request by federicobadini**
closes #729 

### QA

1. Visit Vercel preview
2. You can check a product like [iopener](https://react-commerce-git-add-shipping-restriction-detai-edc61e-ifixit.vercel.app/products/iopener) that has been modified in test to provide both a single and a double restriction 
3. As stated in the document, only the first restriction in the array should be shown. 
4. `fulfilled_via_dropshipping` is the only exception to the rule above: if present it needs always to be shown
5. Check that is working also on mobile

### CODE REVIEW

File: frontend/templates/product/sections/ProductSection/AddToCart/index.tsx [hunk 1]
```diff
@@ -106,19 +122,108 @@ export function AddToCart({ product, selectedVariant }: AddToCartProps) {
...
+   const restrictionKeysToShow: string[] = [];
+   const primaryRestrictionKey = shippingRestrictions.find(
+      (sr) =>
+         sr !== 'fulfilled_via_dropshipping' &&
+         Object.keys(shippingRestrictionsInfo).includes(sr)
    );
+   if (primaryRestrictionKey) {
+      restrictionKeysToShow.push(primaryRestrictionKey);
+   }
+   if (shippingRestrictions?.includes('fulfilled_via_dropshipping')) {
```
Comment (masonmcelvain) [inline frontend/templates/product/sections/ProductSection/AddToCart/index.tsx:169]: `shippingRestrictions` can't be undefined/null here right?

```suggestion
   if (shippingRestrictions.includes('fulfilled_via_dropshipping')) {
```

**Review by masonmcelvain — DISMISSED** [review]
CR ✌🏻 with a question. Nice job!

**Comment by jordycosta** [comment]
QA 🟢 

Shipping restriction details are working on both desktop + mobile 🦖 
`fulfilled_via_dropshipping` is displayed as an exception to only one rule being shown 👍 

hover-1

hover-2

---

deploy_block 🟨 

Should the link inside these hover tips open up in a new tab? _It currently doesn't_
Also, **Link** replaced **Learn More**

File: frontend/templates/product/sections/ProductSection/AddToCart/index.tsx (hunk 1 above)
Comment (federicobadini) [reply frontend/templates/product/sections/ProductSection/AddToCart/index.tsx]: Correct, I've applied the suggested change

**Review by masonmcelvain — DISMISSED** [review]
CR ✌🏻

**Comment by erinemay** [comment]
>For reference, the priority order is:
>
>battery
dangerous
us only
does not ship to mexico
>
>But this shouldn't need to be coded into the frontend.


https://github.com/iFixit/react-commerce/issues/729#issue-1384256168 I'm not seeing this priority being respected. I'm assuming this is another pull to fix it in the backend?

Sample product: https://react-commerce-git-add-shipping-restriction-detai-edc61e-ifixit.vercel.app/products/google-pixel-3a-xl-replacement-battery-genuine
https://ifixit-test.myshopify.com/admin/products/6785031569498/variants/40088217321562
image
image

**Comment by masonmcelvain** [comment]
@erinemay Currently this is the order on the backend. It's easy to change if we want to change it. But yea, no need to block this pull on it.

image

In ProductExportShippingRestrictions.php

**Comment by jordycosta** [comment]
QA 🟢 Carry-over

https://github.com/iFixit/react-commerce/pull/858#issuecomment-1287120040

**Comment by masonmcelvain** [comment]
Looks like this is still deploy blocked, right @jordycosta?

**Comment by jordycosta** [comment]
> Looks like this is still deploy blocked, right @jordycosta?

Yes, these aren't dev_block concerns, but I did want to note two things:

> deploy_block 🟨 on a hyperlink question

Should the link inside these hover tips open up in a new tab? _It currently doesn't_
Also, **Link** replaced **Learn More**

**Comment by jordycosta** [comment]
QA 🟢 

Link -> Learn More (opens in a new tab) ✔️

**Review by masonmcelvain — DISMISSED** [review]
CR ✌🏻 I think the e2e failure is due to Strapi having issues, and unrelated to this PR. However, it looks like there is a merge conflict.

**Comment by masonmcelvain** [comment]
CR ✌🏻 QA ✌🏻 carry over. Straightforward merge conflict in some generated types.

**Review by masonmcelvain — APPROVED** [review]

**Pull request merged**